src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    let nextState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      nextState = { ...nextState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => {
+        delete nextState[key];
+      });
+    }
+
+    if (action.type === 'clear') {
+      nextState = {};
+    }
```

Comment:
[CODE STYLE]: Instead of using multiple `if` statements, consider using a `switch` statement. This will make your code more readable and efficient, especially as the number of `action.type` options increases.